keystorage/aws/src/test/java/tech/pegasys/signers/aws/AwsSecretsManagerTest.java

```diff
@@ -71,255 +76,294 @@ private void verifyEnvironmentVariables() {
+    testSecretsManagerClient =
+        SecretsManagerAsyncClient.builder()
+            .credentialsProvider(credentialsProvider)
+            .region(Region.of(AWS_REGION))
+            .build();
+  }
+
+  private static void initTestVariables() {
+    testSecretNames = new ArrayList<>();
+    allTestSecretTags = new HashMap<>();
+    testSecretSingleTag = new HashMap<>();
+    testSecretMultipleTags = new HashMap<>();
+    testSecretSingleSharedTag = new HashMap<>();
+    testSecretMultipleSharedTags = new HashMap<>();
+  }
+
+  private static void closeTestSecretsManager() {
+    testSecretsManagerClient.close();
+  }
+
+  private static void closeAwsSecretsManagers() {
+    awsSecretsManagerDefault.close();
+    awsSecretsManagerExplicit.close();
+    awsSecretsManagerInvalidCredentials.close();
+  }
+
+  private static void closeClients() {
+    closeAwsSecretsManagers();
+    closeTestSecretsManager();
+  }
+
+  private static void createSecret(final boolean multipleTags, final boolean sharedTags) {
+    testSecretNamePrefix = "signers-aws-integration/";
+    testSecretName = testSecretNamePrefix + UUID.randomUUID();
+
+    final List<Tag> tags = new ArrayList<>();
+    tags.add(createTag(testSecretName, testSecretNamePrefix));
+
+    if (multipleTags) {
+      tags.add(
+          createTag(
+              testSecretNamePrefix + UUID.randomUUID(), testSecretNamePrefix + UUID.randomUUID()));
+    }
+    if (sharedTags) {
+      allTestSecretTags.forEach((key, value) -> tags.add(createTag(key, value)));
+    }
+
+    final CreateSecretRequest secretRequest =
+        CreateSecretRequest.builder()
+            .name(testSecretName)
+            .secretString(SECRET_VALUE)
+            .tags(tags)
+            .build();
+
+    testSecretsManagerClient.createSecret(secretRequest).join();
+    testSecretNames.add(testSecretName);
+    updateTestTags(tags, multipleTags, sharedTags);
+    waitUntilSecretAvailable(testSecretName);
+  }
+
+  private static void createTestSecrets() {
+    createSecret(false, false);
+    createSecret(true, false);
+    createSecret(false, true);
+    createSecret(true, true);
+  }
+
+  private static void deleteSecrets() {
+    testSecretNames.forEach(
+        name -> {
+          final DeleteSecretRequest deleteSecretRequest =
+              DeleteSecretRequest.builder().secretId(name).build();
+          testSecretsManagerClient.deleteSecret(deleteSecretRequest).join();
+        });
+    testSecretNames.clear();
+    allTestSecretTags.clear();
+    testSecretSingleTag.clear();
+    testSecretMultipleTags.clear();
+    testSecretSingleSharedTag.clear();
+    testSecretMultipleSharedTags.clear();
+  }
+
+  private static Tag createTag(final String key, final String value) {
+    return Tag.builder().key(key).value(value).build();
+  }
+
+  private static void updateTestTags(
+      final List<Tag> tags, final boolean multipleTags, final boolean sharedTags) {
+    tags.forEach(
```

Comment:
nit: find this a bit confusing. Think it would be easier to follow if we just created the sets of tags upfront and then pass them to the secret we want to be created instead.

Reply:
Agree it would be easier to set up some tags first, then then match them up with the secrets as required.